src/database/post.ts

```diff
@@ -3,92 +3,155 @@ import { PostIdentifier, ProfileIdentifier, Identifier, PostIVIdentifier, GroupI
-                    for (const [key, value] of Object.entries(oldType.recipients)) {
-                        newType.set(key, value)
+                /**
+                 * In the version 3 we use `recipients?: Record<string, RecipientDetail>`
+                 * After upgrade to version 4, we use `recipients: IdentifierMap<ProfileIdentifier, RecipientDetail>`
+                 */
+                if (oldVersion <= 3) {
+                    const store = transaction.objectStore('post')
+                    for await (const cursor of store) {
+                        const v3Record: Version3PostRecord = cursor.value as any
+                        const newType: Version4PostRecord['recipients'] = new Map()
+                        for (const [key, value] of Object.entries(v3Record.recipients)) {
+                            newType.set(key, value)
+                        }
+                        const v4Record: Version4PostRecord = {
+                            ...v3Record,
+                            recipients: newType,
+                        }
+                        await cursor.update(v4Record as any)
                     }
-                    const next: PostDBRecord = {
-                        ...cursor.value,
-                        recipients: newType,
+                }
+                /**
+                 * In version 4 we use CryptoKey, in version 5 we use JsonWebKey
+                 */
+                if (oldVersion <= 4) {
+                    const store = transaction.objectStore('post')
+                    for await (const cursor of store) {
+                        const v4Record: Version4PostRecord = cursor.value as any
+                        const data = knowledge?.data!
+                        if (!v4Record.postCryptoKey) continue
+                        const v5Record: Version5PostRecord = {
+                            ...v4Record,
+                            postCryptoKey: data.get(v4Record.identifier)!,
+                        }
+                        if (!v5Record.postCryptoKey) delete v5Record.postCryptoKey
+                        await cursor.update(v5Record as any)
                     }
-                    cursor.update(next)
                 }
+            },
+        }),
+    async (db): Promise<UpgradeKnowledge> => {
+        if (db.version === 4) {
+            const map = new Map<string, AESJsonWebKey>()
+            const knowledge: UpgradeKnowledge = { version: 4, data: map }
+            const records = await createTransaction(db, 'readonly')('post').objectStore('post').getAll()
+            for (const r of records) {
+                const x = r.postCryptoKey
+                if (!x) continue
+                const key = await CryptoKeyToJsonWebKey<AESJsonWebKey>(x as any).catch(() => undefined)
+                if (!key) continue
+                map.set(r.identifier, key)
             }
-        },
-    })
-})
+            return knowledge
+        }
+        return undefined
```

Comment:
Remove it.

Reply:
Not all code paths return a value.ts(7030)